spring/spring5/spring5-embedded/src/main/java/org/infinispan/spring/embedded/SpringEmbeddedModule.java

```diff
@@ -0,0 +1,29 @@
+package org.infinispan.spring.embedded;
+
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.factories.annotations.InfinispanModule;
+import org.infinispan.lifecycle.ModuleLifecycle;
+import org.infinispan.marshall.protostream.impl.SerializationContextRegistry;
+import org.infinispan.spring.common.provider.NullValue;
+import org.infinispan.spring.common.provider.PersistenceContextInitializerImpl;
+
+/**
+ * Add support for Spring-specific classes like {@link NullValue} in embedded caches.
+ *
+ * @author Dan Berindei
+ * @since 12.1
+ */
+@InfinispanModule(name = "spring-embedded", requiredModules = "core")
+public class SpringEmbeddedModule implements ModuleLifecycle {
+
+   @Override
+   public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguration globalConfiguration) {
+      SerializationContextRegistry ctxRegistry = gcr.getComponent(SerializationContextRegistry.class);
+      PersistenceContextInitializerImpl sci = new PersistenceContextInitializerImpl();
+      ctxRegistry.addContextInitializer(SerializationContextRegistry.MarshallerType.PERSISTENCE, sci);
+      ctxRegistry.addContextInitializer(SerializationContextRegistry.MarshallerType.GLOBAL, sci);
```

Comment:
Does `NullValue` need to be persisted?

If so, it should be sufficient for the SCI to just be registered with the PERSISTENCE context.

Reply:
@ryanemerson I tried adding the SCI only to the PERSISTENCE context and it didn't work.
I didn't dig too deep into why it didn't work, but I noticed the same pattern in several other places:

https://github.com/infinispan/infinispan/blob/d8e55acc486c885c87483efefa02b15425260549/remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/LifecycleManager.java#L79-L80
https://github.com/infinispan/infinispan/blob/ecb0ff73fbc8df9898c96a79c4c300d44c239c0c/server/memcached/src/main/java/org/infinispan/server/memcached/LifecycleCallbacks.java#L24-L25

Reply:
@danberindei Can you elaborate on "it didn't work"? Was it a specific aspect of marshalling, e.g. transcoding, cluster communication etc.

Reply:
@ryanemerson like I said, I didn't dig deep because I had already copied both lines, I just commented out the GLOBAL line to see if it worked and uncommented it when I saw that it didn't. 

If I comment the GLOBAL line again, I get this error pointing to the transcoder:

```
[OK:    47, KO:     1, SKIP:     0] Test failed: org.infinispan.spring.embedded.provider.SpringCacheTest.testCacheContainsSupportsNullValue[application/x-protostream]
org.infinispan.commons.marshall.MarshallingException: ISPN000615: Unable to unmarshall 'org.infinispan.spring.common.provider.NullValue' as a marshaller is not present in the user or global SerializationContext
	at org.infinispan.encoding.ProtostreamTranscoder.getCtxForMarshalling(ProtostreamTranscoder.java:198)
	at org.infinispan.encoding.ProtostreamTranscoder.marshall(ProtostreamTranscoder.java:127)
	at org.infinispan.encoding.ProtostreamTranscoder.transcode(ProtostreamTranscoder.java:68)
	at org.infinispan.encoding.DataConversion.toStorage(DataConversion.java:223)
	at org.infinispan.cache.impl.EncoderCache.valueToStorage(EncoderCache.java:107)
	at org.infinispan.cache.impl.EncoderCache.put(EncoderCache.java:715)
	at org.infinispan.spring.common.provider.SpringCache.put(SpringCache.java:142)
	at org.infinispan.spring.embedded.provider.SpringCacheTest.testCacheContainsSupportsNullValue(SpringCacheTest.java:136)
```

Reply:
Thanks for checking. This is an oversight on my behalf. Leave your code as is and I'll address in [ISPN-12696](https://issues.redhat.com/browse/ISPN-12696)

Reply:
Thanks @ryanemerson !